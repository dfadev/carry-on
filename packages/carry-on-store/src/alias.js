import { forceArray, isString } from "carry-on-utils";
import { register } from "./store";

export const state = (init, id, path, dispose) => {
  // storeId can be the first parameter
  if (isString(init)) {
    const actualId = init;
    init = id;
    id = actualId;
  }

  const inits = forceArray(init).map(item => ({
    state: item,
    dispose
  }));

  register(inits, id, path);
};

export const middleware = (init, id, path, dispose) => {
  // storeId can be the first parameter
  if (isString(init)) {
    const actualId = init;
    init = id;
    id = actualId;
  }

  const inits = forceArray(init).map(item => ({
    middleware: item,
    dispose
  }));

  register(inits, id, path);
};
