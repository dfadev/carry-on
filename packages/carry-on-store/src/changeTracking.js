/** @format **/
import { keys, proxyState, deproxify, mutateSetA } from "carry-on-utils";

export function compareChanges(changes, affected) {
  const queue = [];
  queue.push({ changes, affected });

  while (queue.length > 0) {
    const item = queue.pop();
    const entries = item.changes;

    for (let i = 0, len = entries.length; i < len; i++) {
      const entry = entries[i];
      const key = entry.key;
      const affectedValue = item.affected[key];

      if (affectedValue === true) return true;
      if (affectedValue !== undefined) {
        const nextChanges = entry.changes;

        if (nextChanges === true) return true;
        queue.push({
          changes: nextChanges,
          affected: affectedValue
        });
      }
    }
  }

  return false;
}

export function trackChanges(state, select) {
  const trappedState = proxyState(state);
  const selectedState = select(trappedState.state);
  trappedState.seal();
  const affected = trappedState.affected;
  const deproxified = deproxify(selectedState);
  const finalState = deproxified !== undefined ? deproxified : selectedState;

  return {
    finalState,
    affected
  };
}

export function calculateChanges(patches) {
  const stage1 = {};
  for (let i = 0, len = patches.length; i < len; i++)
    mutateSetA(stage1, patches[i].path, true);

  // precompute object walk so keys is called the least amount necessary
  const stage2 = [];
  const queue = [];
  queue.push({ keys: keys(stage1), changes: stage1, out: stage2 });

  while (queue.length > 0) {
    const item = queue.pop();

    for (let i = 0, len = item.keys.length; i < len; i++) {
      const key = item.keys[i];
      const changes = item.changes[key];
      if (changes === true) {
        item.out.push({ key, changes });
      } else {
        const nextChanges = [];
        queue.push({
          keys: keys(changes),
          changes,
          out: nextChanges
        });
        item.out.push({ key, changes: nextChanges });
      }
    }
  }

  return stage2;
}
