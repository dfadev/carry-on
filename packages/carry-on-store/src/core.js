/** @format **/
import { isFunction } from "./utils";

export default function makeStoreModule(defaultId, extra = () => ({})) {
  // wrap a function with middleware
  const applyMiddleware = (middlewares, fn, apply) => {
    if (!Array.isArray(middlewares)) middlewares = [middlewares];
    for (const middleware of middlewares) fn = apply(middleware, fn);
    return (...args) => fn(...args);
  };

  // middleware initialize message type
  const initMessage = "Initialize";

  const createPlugins = (store, plugins) => {
    if (!plugins) return;

    // create plugins
    store.plug = {};

    if (!Array.isArray(plugins)) plugins = [plugins];

    for (const plugin of plugins) {
      const { id, state, dispatch } = plugin;
      if (state)
        store.plug[id] = isFunction(state)
          ? state(store.dispatch, store.query, store.plug)
          : state;
      if (dispatch)
        store.d = applyMiddleware(dispatch, store.d, (middleware, fn) =>
          middleware(fn, store.query, store.plug)
        );
    }
  };

  // a map of stores
  const stores = {};

  // create a store
  function create(id) {
    return Object.assign({ id, pending: [] }, extra());
  }

  // delete a store
  function deleteStore(id = defaultId) {
    delete stores[id];
  }

  // lookup a store
  const useStore = (id = defaultId) => stores[id] || (stores[id] = create(id));

  // register state
  const register = (init, id = defaultId) => {
    const { dispatch, query, pending } = useStore(id);
    // queue if no dispatch available yet
    if (dispatch)
      dispatch(state =>
        Object.assign(state, isFunction(init) ? init(dispatch, query) : init)
      );
    else pending.push(init);
  };

  const connect = ({ id, producer, publish, plugins, init }) => {
    const store = useStore(id);
    if (store.dispatch) throw new Error("Already connected");

    // the producer creates new states from old states by executing actions
    store.producer = producer || ((state, action) => ({ ...action(state) }));

    // query provides a copy of state created by the producer
    store.query = (action = identity => identity, ...args) =>
      store.producer(store.state, action, ...args);

    // run producer action and set state
    store.d = (action, type, force) => {
      const nextState = (store.state = force
        ? action(store.state)
        : store.producer(store.state, action));
      if (type !== initMessage) {
        publish(nextState);
      }
      return nextState;
    };
    store.dispatch = (...args) => store.d(...args);

    createPlugins(store, plugins);
    store.dispatch = store.d;

    // create initial state
    store.state = Object.assign(
      {},
      // execute store initialization
      isFunction(init) ? init(store.dispatch, store.query, store.plug) : init,
      // execute pending store initialization
      ...store.pending.map(pending =>
        pending(store.dispatch, store.query, store.plug)
      ),
      // add standard keys
      {
        id: store.id,
        dispatch: store.dispatch,
        query: store.query,
        plug: store.plug
      }
    );

    // initialize middleware with state
    if (plugins) store.dispatch(() => store.state, initMessage);

    return store.state;
  };

  return {
    useStore,
    deleteStore,
    connect,
    defaultId,
    register
  };
}
