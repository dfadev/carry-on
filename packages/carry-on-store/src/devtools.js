/** @format **/
export default function devTools({ timeTravel = true } = {}) {
  // dev tools connections
  const connections = [],
    subscriptions = [],
    // time travel state tracking (only for immutable state)
    time = {},
    // check for dev tools extension
    devToolsExt = window && window.__REDUX_DEVTOOLS_EXTENSION__;

  if (!devToolsExt) return {};

  return {
    middleware: ({ dispatch, id }) =>
      function devToolsMiddleware(action, type = "Dispatch", ...args) {
        const state = dispatch(action, type, ...args);

        // exit when no action
        if (!action) return state;
        const name = id;

        let connection = connections[name];
        if (!connection)
          connection = connections[name] = devToolsExt.connect({ name });

        // support time traveling
        if (timeTravel) {
          const states = time[name] || (time[name] = []);
          states.push(state);
          subscriptions[name] ||
            (subscriptions[name] = connection.subscribe(
              msg =>
                msg.type === "DISPATCH" &&
                msg.payload &&
                msg.payload.type === "JUMP_TO_STATE" &&
                dispatch(s => {
                  if (s === undefined) s = {};
                  const keys = Object.keys(s);
                  for (let i = 0, len = keys.length; i < len; i++)
                    delete s[keys[i]];
                  return Object.assign(s, states[msg.payload.index]);
                }, "Time Travel")
            ));
        }

        // send devtools an state update message
        connection.send({ type }, state);

        return state;
      }
  };
}
