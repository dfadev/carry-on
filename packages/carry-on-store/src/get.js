import { isString } from "carry-on-utils";
import { getStore } from "./store";

export default function get(fn, id, opts) {
  if (isString(fn)) {
    const actualId = fn;
    fn = id;
    id = actualId;
  }

  const store = getStore(id);
  return store.get(fn, opts);
}
