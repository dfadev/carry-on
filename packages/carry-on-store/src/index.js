/** @format **/
export default function makeStoreModule(defaultId, extra = () => ({})) {
  const isFunction = thing =>
    !!(thing && thing.constructor && thing.call && thing.apply);

  // wrap a function with middleware
  const applyMiddleware = (middlewares, fn, apply) => {
    if (!Array.isArray(middlewares)) middlewares = [middlewares];
    for (const middleware of middlewares) fn = apply(middleware, fn);
    return (...args) => fn(...args);
  };

  // middleware initialize message type
  const initMessage = "Initialize";

  const createPlugins = (store, plugins) => {
    // create plugins
    store.plug = {};

    if (!Array.isArray(plugins)) plugins = [plugins];

    // dispatch
    for (const plugin of plugins) {
      const { dispatch } = plugin;
      if (dispatch)
        store.d = applyMiddleware(dispatch, store.d, (middleware, fn) =>
          middleware(fn, store.query, store.plug)
        );
    }

    // state
    for (const plugin of plugins) {
      const { id, state } = plugin;
      if (state)
        store.plug[id] = isFunction(state)
          ? state(store.d, store.query, store.plug)
          : state;
    }
  };

  // a map of stores
  const stores = {};

  // create a store
  function create(id) {
    return Object.assign({ id, pending: [] }, extra());
  }

  // delete a store
  function deleteStore(id = defaultId) {
    delete stores[id];
  }

  // lookup a store
  const useStore = (id = defaultId) => stores[id] || (stores[id] = create(id));

  // register state
  const register = (init, id = defaultId) => {
    const { dispatch, query, pending } = useStore(id);
    // queue if no dispatch available yet
    if (dispatch)
      dispatch(state =>
        Object.assign(
          {},
          state,
          isFunction(init) ? init(dispatch, query) : init
        )
      );
    else pending.push(init);
  };

  const connect = ({ id, producer, publish, plugins, init } = {}) => {
    const store = useStore(id);
    if (store.dispatch) throw new Error("Already connected");

    // the producer creates new states from old states by executing actions
    store.producer = producer || ((state, action) => action({ ...state }));

    // query provides a copy of state created by the producer
    store.query = (action = identity => identity, ...args) =>
      store.producer(store.state, action, ...args);

    // run producer action and set state
    store.d = (action, type, force) => {
      const nextState = (store.state = force
        ? action(store.state)
        : store.producer(store.state, action));
      if (publish && type !== initMessage) {
        publish(nextState);
      }
      return nextState;
    };

    if (plugins) createPlugins(store, plugins);
    store.dispatch = store.d;

    // create initial state
    store.state = Object.assign(
      {},
      // execute store initialization
      isFunction(init) ? init(store.dispatch, store.query, store.plug) : init,
      // execute pending store initialization
      ...store.pending.map(
        pending =>
          isFunction(pending)
            ? pending(store.dispatch, store.query, store.plug)
            : pending
      ),
      // add standard keys
      {
        dispatch: store.dispatch,
        query: store.query,
        plug: store.plug
      },
      store.id ? { id: store.id } : undefined
    );

    // initialize middleware with state
    if (plugins) store.dispatch(() => store.state, initMessage);

    return store.state;
  };

  return {
    useStore,
    deleteStore,
    connect,
    defaultId,
    register
  };
}
