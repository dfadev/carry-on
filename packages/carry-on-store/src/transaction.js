import { keys } from "carry-on-utils";

export default function transaction() {
  const transactions = [];

  return {
    state: ({ set, get }) => ({
      commit: () => {
        if (transactions.length === 0) throw new Error("no tx");
        return set(state => {
          transactions.pop();
          return state;
        }, "Commit");
      },
      rollback: () => {
        if (transactions.length === 0) throw new Error("no tx");
        return transactions.pop()();
      },
      begin: () =>
        set(state => {
          const rollbackState = get();
          transactions.push(() =>
            set(s => {
              const keyList = keys(s);
              for (let i = 0, len = keyList.length; i < len; i += 1)
                delete s[keyList[i]];
              const newState = Object.assign(s, rollbackState);
              return newState;
            }, "Rollback")
          );
          return state;
        }, "Begin Transaction")
    }),
    dispose: () => {
      transactions.length = 0;
    }
  };
}
