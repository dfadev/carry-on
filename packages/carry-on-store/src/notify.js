import { compareChanges } from "./changeTracking";

export default function notify() {
  const subscribers = new Map();

  function subscribe(fn, watch) {
    subscribers.set(fn, watch);
    return function unsusbscribe() {
      subscribers.delete(fn);
    };
  }

  const plugin = {
    middleware: ({ next, getChanges, wrap, isNested }) =>
      function notifyMiddleware(action, type, opts, ...args) {
        const state = next(action, type, opts, ...args);
        if (isNested()) return state; // don't notify on nested sets

        const changes = getChanges && getChanges();

        const notifySubs = () => {
          const notifyThese = [];
          const entries = subscribers.entries();
          let entry = entries.next();
          while (!entry.done) {
            const [fn, watch] = entry.value;
            if (watch === undefined) notifyThese.push(fn);
            else {
              const hasChanges = compareChanges(changes, watch);
              if (hasChanges) notifyThese.push(fn);
            }
            entry = entries.next();
          }

          for (let i = 0, len = notifyThese.length; i < len; i += 1)
            notifyThese[i](state, changes);
        };

        if (opts && opts.immediate) notifySubs();
        else wrap(notifySubs);

        return state;
      },
    dispose: () => {
      subscribers.clear();
    }
  };

  return {
    subscribers,
    subscribe,
    plugin
  };
}
