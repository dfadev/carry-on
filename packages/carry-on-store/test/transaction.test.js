import transaction from "../src/transaction";
import immer from "immer";
import { deleteStore, register, getStore, connect } from "../src";

test("transaction match", () => {
  expect(transaction()).toMatchSnapshot();
});

test("plug snapshot", () => {
  const plug = {};
  let state = { plug };
  const get = () => state;
  const set = (action, type) => {
    return (state = action(state));
  };

  plug.tx = transaction().state({ set, get, plug });

  expect(plug.tx).toMatchSnapshot();
});

test("commit", () => {
  const get = () => {};
  const plug = {};
  let state = { plug };
  const set = (action, type) => {
    return (state = action(state));
  };

  plug.tx = transaction().state({ set, get, plug });

  const beginState = plug.tx.begin();

  const newState = set(state => {
    return { ...state, newThing: 1 };
  });
  expect(newState.newThing).toBe(1);

  const commitedState = plug.tx.commit();

  expect(commitedState).toBe(newState);
  expect(commitedState.newThing).toEqual(1);
});

test("rollback", async () => {
  const get = state => immer(state, s => s);
  const state = {};
  const set = (action, type) => {
    return immer(state, action);
  };

  const store = {
    set: (action, type) => {
      return (store.state = immer(store.state, action));
    },
    get,
    state: transaction().state({ set, get })
  };

  const beginState = store.state.begin();

  const newState = store.set(state => {
    state.newThing = 1;
  });

  const rolledBackState = store.state.rollback();

  expect(rolledBackState).toBe(beginState);
  expect(rolledBackState.newThing).toBe(undefined);
});

test("commit no transaction throws", () => {
  const get = () => {};
  const plug = {};
  let state = { plug };
  const set = (action, type) => {
    return (state = action(state));
  };

  plug.tx = transaction().state({ set, get, plug });

  expect(plug.tx.commit).toThrow();
});

test("rollback no transaction throws", () => {
  const get = () => {};
  const plug = {};
  let state = { plug };
  const set = (action, type) => {
    return (state = action(state));
  };

  plug.tx = transaction().state({ set, get, plug });

  expect(plug.tx.rollback).toThrow();
});

test("dispose", () => {
  transaction().dispose();
});

test("commit 2", () => {
  const store = getStore();
  register(transaction());
  connect();
  store.get().begin();
  store.set(state => {
    state.extra = 1;
  });
  expect(store.get()).toMatchSnapshot();
  store.get().rollback();
  expect(store.get()).toMatchSnapshot();
  deleteStore();
});
