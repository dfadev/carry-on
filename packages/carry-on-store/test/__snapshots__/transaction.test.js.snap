// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commit 2 1`] = `
Object {
  "begin": [Function],
  "commit": [Function],
  "extra": 1,
  "rollback": [Function],
}
`;

exports[`commit 2 2`] = `
Object {
  "begin": [Function],
  "commit": [Function],
  "rollback": [Function],
}
`;

exports[`plug snapshot 1`] = `
Object {
  "begin": [Function],
  "commit": [Function],
  "rollback": [Function],
}
`;

exports[`transaction match 1`] = `
Object {
  "dispose": [Function],
  "state": [Function],
}
`;
